src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -45,19 +45,173 @@ struct extent_to_write_t {
   laddr_t addr = L_ADDR_NULL;
   extent_len_t len;
   std::optional<bufferlist> to_write;
+  std::optional<paddr_t> paddr;
 
   extent_to_write_t() = default;
   extent_to_write_t(const extent_to_write_t &) = default;
   extent_to_write_t(extent_to_write_t &&) = default;
 
   extent_to_write_t(laddr_t addr, bufferlist to_write)
-    : addr(addr), len(to_write.length()), to_write(to_write) {}
+    : addr(addr), len(to_write.length()), to_write(to_write), paddr(std::nullopt) {}
 
   extent_to_write_t(laddr_t addr, extent_len_t len)
-    : addr(addr), len(len) {}
+    : addr(addr), len(len), to_write(std::nullopt), paddr(std::nullopt) {}
+
+  extent_to_write_t(laddr_t addr, paddr_t paddr, extent_len_t len)
+    : addr(addr), len(len), to_write(std::nullopt), paddr(paddr) {}
 };
 using extent_to_write_list_t = std::list<extent_to_write_t>;
 
+struct split_info_t {
+  laddr_t pin_begin;
+  laddr_t pin_end;
+  paddr_t left_paddr;
+  paddr_t right_paddr;
+  laddr_t data_begin;
+  laddr_t data_end;
+  laddr_t aligned_begin;
+  laddr_t aligned_end;
+  extent_len_t block_size;
+  extent_len_t data_size;
+  extent_len_t write_size;
+  bool split_left;
+  bool split_right;
+  friend std::ostream& operator<<(std::ostream& out, const split_info_t& info) {
+    return out << "pin_begin " << info.pin_begin
+	       << " pin_end " << info.pin_end
+	       << " left_paddr " << info.left_paddr
+	       << " right_paddr " << info.right_paddr
+	       << " data_begin " << info.data_begin
+	       << " data_end " << info.data_end
+	       << " aligned_begin " << info.aligned_begin
+	       << " aligned_end " << info.aligned_end
+	       << " block_size " << info.block_size
+	       << " data_size " << info.data_size
+	       << " write_size " << info.write_size << std::boolalpha
+	       << " split_left " << info.split_left
+	       << " split_right " << info.split_right << std::noboolalpha;
+  }
+  void init(laddr_t offset,
+	    extent_len_t len,
+	    const lba_pin_list_t& pins,
+	    extent_len_t _block_size) {
```

Comment:
Move init() to be a free function which returns a overwrite_plan_t:
```
overwrite_plan_t generate_overwrite_plan(
    laddr_t offset,
    extent_len_t len,
    const lba_pin_list_t& pins,
    extent_len_t _block_size)
```

Reply:
This is complicated enough to deserve dedicated unit tests, add a test specifically for generate_overwrite_plan to test_object_data_handler.cc.